File: nodejs-mvc/hexagonal-architecture/src/domain/port/output/ProductRepositoryPort.ts
import { Product } from '../../model/Product';

/**
 * @layer Domain - Output Port
 * @description Defines the operations that the domain needs from the outside world.
 * This is the SECONDARY/DRIVEN port in Hexagonal Architecture.
 *
 * Output ports define WHAT the domain needs from persistence, without
 * specifying HOW it's implemented (in-memory, SQL, NoSQL, etc.)
 * The domain service depends on this interface, NOT on concrete implementations.
 */
export interface ProductRepositoryPort {
  save(product: Product): Promise<Product>;
  findById(id: string): Promise<Product | null>;
  findAll(page: number, size: number): Promise<{ products: Product[]; total: number }>;
  findBySku(sku: string): Promise<Product | null>;
  search(query?: string, category?: string, minPrice?: number, maxPrice?: number): Promise<Product[]>;
  update(product: Product): Promise<Product>;
  delete(id: string): Promise<void>;
}

File: python-mvc/hexagonal-architecture/app/domain/port/output/product_repository_port.py
"""
Output Port: Product Repository Port.

This port defines what the domain NEEDS from the persistence layer.
It is part of the domain and must not depend on any infrastructure detail.

In Hexagonal Architecture, output ports are implemented by output adapters
(e.g., database repositories, external API clients, file system adapters).
The domain depends on this abstraction, not on concrete implementations.
"""
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Optional

from app.domain.model.product import Product


class ProductRepositoryPort(ABC):
    """
    Output port for product persistence operations.

    This interface is defined in the domain layer and implemented
    by output adapters (InMemoryRepository, PostgresRepository, etc.).
    """

    @abstractmethod
    async def save(self, product: Product) -> Product:
        """Persist a new product."""
        ...

    @abstractmethod
    async def find_by_id(self, product_id: str) -> Optional[Product]:
        """Find a product by its unique identifier."""
        ...

    @abstractmethod
    async def find_all(self, page: int, size: int) -> tuple[list[Product], int]:
        """
        Find all products with pagination.

        Returns a tuple of (products_list, total_count).
        """
        ...

    @abstractmethod
    async def find_by_sku(self, sku: str) -> Optional[Product]:
        """Find a product by its SKU."""
        ...

    @abstractmethod
    async def search(
        self,
        query: Optional[str] = None,
        category: Optional[str] = None,
        min_price: Optional[float] = None,
        max_price: Optional[float] = None,
    ) -> list[Product]:
        """Search products by various criteria."""
        ...

    @abstractmethod
    async def update(self, product: Product) -> Product:
        """Update an existing product."""
        ...

    @abstractmethod
    async def delete(self, product_id: str) -> None:
        """Delete a product by its identifier."""
        ...
